epitator/structured_incident_annotator.py

```diff
@@ -0,0 +1,287 @@
+        }
+
+        tables = []
+        for span in doc.tiers['structured_data'].spans:
+            if span.metadata['type'] != 'table': continue
+            # Add columns based on surrounding text
+            __section_title = doc.tiers['spacy.sentences'].nearest_prior_span(span)
+            __last_geoname_mentioned = geonames.nearest_prior_span(span)
+            __last_date_mentioned = dates.nearest_prior_span(span)
+            rows = span.metadata['data']
+            # Detect header
+            all_entities = [value
+                for value_group in entities_by_type.values()
+                for value in value_group]
+            first_row = AnnoTier(rows[0])
+            header_entities = list(first_row.group_spans_by_containing_span(numbers))
+            if all(len(entity_spans) == 0 for header_span, entity_spans in header_entities):
+                has_header = True
+            else:
+                has_header = False
+            if has_header:
+                data_rows = rows[1:]
+            else:
+                data_rows = rows
+
+            # Determine column types
+            table_by_column = zip(*data_rows)
+            column_types = []
+            parsed_column_entities = []
+            for column_values in table_by_column:
+                num_non_null_rows = sum(not is_null(value.text) for value in column_values)
+                column_values = AnnoTier(column_values)
+                # Choose column type based on greatest percent match,
+                # if under 30, choose text.
+                max_matches = 0
+                matching_values = None
+                matching_column_entities = None
+                column_type = "text"
+                for value_type, value_spans in entities_by_type.items():
+                    filtered_value_spans = value_spans
+                    if value_type == "number":
+                        filtered_value_spans = value_spans.without_overlaps(dates)
+                    column_entities = [
+                        SpanGroup(contained_spans) if len(contained_spans) > 0 else None
+                        for group_span, contained_spans in column_values.group_spans_by_containing_span(filtered_value_spans)]
+                    num_matches = sum(
+                        contained_spans is not None
+                        for contained_spans in column_entities)
+                    
+                    if num_non_null_rows > 0 and float(num_matches) / num_non_null_rows > 0.3:
+                        if num_matches > max_matches:
+                            max_matches = num_matches
+                            matching_values = value_spans
+                            matching_column_entities = column_entities
+                            column_type = value_type
+                    if matching_column_entities is None:
+                        matching_column_entities = [[] for x in column_values]
+                column_types.append(column_type)
```

Comment:
Might it be better to rely on something like `pandas` for value inference? Like, get an extracted column and shove it into a `DataFrame`? I'm guessing your answer would be "no" because that adds dependencies and whatnot?

Reply:
No so much because of the dependencies, but rather I think we can do a better job of value inference than pandas in this context by customizing things to the type of data we're extracting.

Reply:
Yeah, that makes sense